certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,378 @@
+        (ghostGuardians[wallet][SENTINEL] == NULL || ghostGuardians[wallet][SENTINEL] == SENTINEL)
+    {
+        preserved {
+            requireInvariant reach_invariant();
+            requireInvariant reach_null();
+            requireInvariant inListReachable();
+            requireInvariant reach_next();
+            requireInvariant nextNull();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+            requireInvariant count_correct();
+            requireInvariant guardiancount_correct();
+        }
+    }
+
+invariant emptyListNotReachable()
+    forall address wallet. (ghostGuardians[wallet][SENTINEL] == NULL || ghostGuardians[wallet][SENTINEL] == SENTINEL)
+        => (forall address X. X != SENTINEL => ghostGuardians[wallet][X] == NULL)
+    {
+        preserved {
+            requireInvariant reach_invariant();
+            requireInvariant reach_null();
+            requireInvariant inListReachable();
+            requireInvariant reach_next();
+            requireInvariant nextNull();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+            requireInvariant count_correct();
+            requireInvariant guardiancount_correct();
+        }
+    }
+
+rule storeHookPreservesInvariants(address wallet, address key, address value) {
+    // These are checked in the hook.
+    require key != NULL;
+    require reach(wallet, value, key) => value == SENTINEL; //, "list is cyclic";
+
+    // Invariants that hold even in the middle
+    requireInvariant reach_null();
+    requireInvariant reach_invariant();
+
+    address someKey;
+    address someWallet;
+    require reach_succ(someWallet, someKey, ghostGuardians[someWallet][someKey]);
+    require ghostSuccCount(someWallet, someKey) == count_expected(someWallet, someKey);
+    ghostGuardians[wallet][key] = value;
+    havoc reach assuming updateSucc(wallet, key, next_or_null(value));
+    mathint countDiff = count_expected(wallet, key) - ghostSuccCount(wallet, key);
+    havoc ghostSuccCount assuming updateGhostSuccCount(wallet, key, countDiff);
+    assert reach_succ(someWallet, someKey, ghostGuardians[someWallet][someKey]), "reach_succ violated after guardians update";
+    assert ghostSuccCount(someWallet, someKey) == count_expected(someWallet, someKey);
+
+    // assert also the invariants used above
+    assert (forall address W. forall address X. reach(W, X, NULL));
+    assert (forall address W. forall address X. forall address Y. forall address Z.
+        reach(W, X, X)
+        && (reach(W, X, Y) && reach(W, Y, X) => X == Y)
+        && (reach(W, X, Y) && reach(W, Y, Z) => reach(W, X, Z))
+        && (reach(W, X, Y) && reach(W, X, Z) => (reach(W, Y, Z) || reach(W, Z, Y)))
+    );
+}
+
+rule isGuardianDoesNotRevert {
+    address addr;
+    isGuardian@withrevert(safeContract, addr);
+    assert !lastReverted, "isGuardian should not revert";
+}
+
+rule isGuardianNotSentinal {
```

Comment:
```suggestion
rule isGuardianNotSentinel {
```

Reply:
i'd rename it to `sentinelCantBeGuardian`

Reply:
updated here: https://github.com/5afe/CandideWalletContracts/pull/20/commits/ac730162be230ee32f204453aba53afe8df5fcdd